### PR TITLE
Refactor Material and Shader, Use map unmap

### LINKED ISSUE #31: Update vertex buffers instead of making new ones at runtime
Use ID3D11DeviceContext::Map() and ID3D11DeviceContext::Unmap() to update buffers once they have been built.

### PR DESCRIPTION
**Pull request by sin3point14**
Refactor in MVC and VC is required
closes #31

### CODE REVIEW

File: rootex/core/renderer/material.h [hunk 1]
```diff
@@ -7,40 +7,84 @@ class Texture;
...
 };
 
 class CPUParticlesMaterial : public Material
 {
 public:
+
+	enum class PixelConstantBufferType
+	{
+		Color,
+		END
```
Comment (IronicallySerious) [inline rootex/core/renderer/material.h:83]: End -> End

File: rootex/core/renderer/rendering_device.cpp [hunk 2]
```diff
@@ -349,6 +349,19 @@ void RenderingDevice::bind(ID3D11InputLayout* inputLayout)
...
 }
 
+//Assuming subresource offset = 0
+void RenderingDevice::getBufferMappedContext(ID3D11Buffer* buffer, D3D11_MAPPED_SUBRESOURCE& subresource)
+{
+	m_Context->Map(buffer, 0u, D3D11_MAP_WRITE_DISCARD, 0u, &subresource);
+}
+
+//Assuming subresource offset = 0
+
```
Comment (IronicallySerious) [inline rootex/core/renderer/rendering_device.cpp:359]: Remove newline

File: rootex/core/renderer/rendering_device.h [hunk 3]
```diff
@@ -61,11 +61,11 @@ class RenderingDevice
...
-	void createVSModelConstantBuffer(D3D11_BUFFER_DESC* cbd, D3D11_SUBRESOURCE_DATA* csd);
-	void createVSModelInverseConstantBuffer(D3D11_BUFFER_DESC* cbd, D3D11_SUBRESOURCE_DATA* csd);
-	void createVSViewConstantBuffer(D3D11_BUFFER_DESC* cbd, D3D11_SUBRESOURCE_DATA* csd);
-	void createVSProjectionConstantBuffer(D3D11_BUFFER_DESC* cbd, D3D11_SUBRESOURCE_DATA* csd);
-	void createPSConstantBuffer(D3D11_BUFFER_DESC* cbd, D3D11_SUBRESOURCE_DATA* csd, UINT offset);
+	ID3D11Buffer* createVSModelConstantBuffer(D3D11_BUFFER_DESC* cbd, D3D11_SUBRESOURCE_DATA* csd);
+	ID3D11Buffer* createVSModelInverseConstantBuffer(D3D11_BUFFER_DESC* cbd, D3D11_SUBRESOURCE_DATA* csd);
+	ID3D11Buffer* createVSViewConstantBuffer(D3D11_BUFFER_DESC* cbd, D3D11_SUBRESOURCE_DATA* csd);
+	ID3D11Buffer* createVSProjectionConstantBuffer(D3D11_BUFFER_DESC* cbd, D3D11_SUBRESOURCE_DATA* csd);
+	ID3D11Buffer* createPSConstantBuffer(D3D11_BUFFER_DESC* cbd, D3D11_SUBRESOURCE_DATA* csd, UINT offset);
```
Comment (IronicallySerious) [inline rootex/core/renderer/rendering_device.h:68]: Why not use ComPtr?

File: rootex/core/renderer/shader.h [hunk 4]
```diff
@@ -20,31 +27,23 @@ class Shader
...
 };
 
+class ColorShader : public Shader
+{
+	ColorShader(const LPCWSTR& vertexPath, const LPCWSTR& pixelPath, const BufferFormat& vertexBufferFormat);
+	ColorShader(ColorShader&) = delete;
+	~ColorShader() = default;
+
+	friend class ShaderLibrary;
+
```
Comment (IronicallySerious) [inline rootex/core/renderer/shader.h:45]: Shouldn't there be a set(Color) function here similar to [here](https://github.com/sdslabs/Rootex/pull/136/files#diff-d4a570dc07266a99ae0bd9f70f585311R60)?

File: rootex/framework/components/visual/textured_model_visual_component.cpp [hunk 5]
```diff
@@ -26,15 +26,15 @@ Component* TexturedModelVisualComponent::CreateDefault()
 	ImageResourceFile* imageRes = ResourceLoader::CreateImageResourceFile("rootex/assets/rootex.png");
 	Ref<Texture> texture(new Texture(imageRes));
 
-	Ref<TexturedMaterial> material(new TexturedMaterial(texture));
+	Ref<TextureMaterial> material(new TextureMaterial(texture));
```
Comment (IronicallySerious) [inline rootex/framework/components/visual/textured_model_visual_component.cpp:29]: TextureMaterial -> TexturedMaterial

File: rootex/core/renderer/material.cpp [hunk 6]
```diff
@@ -8,40 +8,192 @@
...
+		m_VSConstantBuffer[(int)VertexConstantBufferType::View] = RenderingDevice::GetSingleton()->createVSViewConstantBuffer(&cbd, &csd);
+		RenderingDevice::GetSingleton()->createVSViewConstantBuffer(&cbd, &csd);
+		break;
+	case VertexConstantBufferType::Projection:
+		m_VSConstantBuffer[(int)VertexConstantBufferType::Projection] = RenderingDevice::GetSingleton()->createVSProjectionConstantBuffer(&cbd, &csd);
+		RenderingDevice::GetSingleton()->createVSProjectionConstantBuffer(&cbd, &csd);
+		break;
+	default:
+		break;
+	}
```
[the remark this answers is not in the document]
Comment (sin3point14) [reply rootex/core/renderer/material.cpp:118]: ye to vs har switch case mein karta hai

File: rootex/core/renderer/shader.h (hunk 4 above)
Comment (sin3point14) [reply rootex/core/renderer/shader.h:45]: Color VC se milta tha pehle and ab MVC store kar tha most probably

File: rootex/core/renderer/material.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (IronicallySerious) [reply rootex/core/renderer/material.cpp:118]: Indentation is wrong at one point at the top

File: rootex/core/renderer/shader.h (hunk 4 above)
Comment (IronicallySerious) [reply rootex/core/renderer/shader.h:45]: But how does it set a color if the shader doesn't have a setColor function?

File: rootex/core/renderer/material.h [hunk 7]
```diff
@@ -7,40 +7,84 @@ class Texture;
...
-	void setShaderConstantBuffer(Shader::VertexConstantBufferType matrixType, const Matrix& matrix) { m_Shader->set(matrixType, matrix); }
-	void setShaderConstantBuffer(PSSolidConstantBuffer& Cb) { m_Shader->set(Cb); }
+	virtual void setVertexShaderConstantBuffer(const VertexConstantBufferType type, const Matrix& constantBuffer);
 };
 
-class TexturedMaterial : public Material
+class ColorMaterial : public Material
+{
+public:
+	enum class PixelConstantBufferType
```
[the remark this answers is not in the document]
Comment (sin3point14) [reply rootex/core/renderer/material.h:41]: yep

File: rootex/core/renderer/shader.h (hunk 4 above)
Comment (sin3point14) [reply rootex/core/renderer/shader.h:45]: Color is set in Constant Buffer so it the relevant functions lie with Material

**Pull request closed without merging**